toolchain/musl: Update to 1.2.5

hnyman opened the pull request:
Update musl C library to 1.2.5

* add `902-basename_hack.patch` to keep the declaration of 'basename'. See explanation below.
* refresh `900-iconv_size_hack.patch`

Nice that musl 1.2.5 was release released well ahead the possible branching of 24.0x, so that there is possibility to have it in main/master for some time before the next release branch.

Compiled for mediatek/filogic MT6000, compiled firmware flashed, runs ok.
Also compiled for ipq806x/generic R7800.

----------------

**Brief explanation of changes from upstream:**

This release adds extension functions statx and preadv2/pwritev2, with fallback implementations for older kernels, and adds two new ports: loongarch64 and riscv32. Minor changes to the printf family of functions have been made for conformance to new standards interpretations/requirements. TLSDESC support for riscv64 has also been added.

Bugs fixed include some DNS issues related to new TCP fallback functionality, several rare race conditions, potentially incorrect return value when glob aborts, and several signifiant arch-specific bugs affecting TLSDESC on arm, riscv64 icache flushing, and sh sigsetjmp and dlsym RTLD_NEXT.

**1.2.5 release notes:**

new features:
- statx function (linux extension; via syscall and fallback using fstatat)
- clone function is now usable and gives _Fork-like consistency in child
- statvfs now provides f_type result
- preadv2 and pwritev2 (linux extension) syscall wrappers
- riscv64 TLSDESC support

new ports:
- loongarch64
- riscv32

compatibility:
- DNS resolver can now handle answers with long CNAME chains
- string.h no longer provides (C23-incompat) non-prototype decl of basename
- fstatat statx backend now matches stat syscall non-automounting behavior
- mntent interfaces now handle escaped whitespace in paths/options

standards updates:
- printf %lc of nul wchar now produces output
- snprintf and swprintf no longer fail on n > INT_MAX
- ppoll is now exposed in default feature profile

bugs fixed:
- some long DNS answers were wrongly rejected despite new TCP support
- glob could wrongly return GLOB_NOMATCH if aborted before any matches
- multithreaded set*id could malfunction from thread sequencing logic bug
- certain use of threads after fork could deadlock thread-list lock
- posix_spawn child could deadlock in race with async parent death
- mbrtowc return value was wrong if argument n exceeded UINT_MAX
- 80-bit extended acoshl and powl got some corner cases wrong
- syslog incorrectly generated localized timestamps

arch-specific bugs fixed:
- arm (32-bit) TLSDESC malfunctioned due to addends being processed wrong
- riscv64 icache flush operation was non-functional
- sh sigsetjmp failed to properly restore call-saved register r8 on return
- sh dlsym RTLD_NEXT did not identify calling module correctly

---------------

**Reasoning for 902-basename_hack.patch:**

Upstream removed the declaration of 'baseline' from string.h with 
https://git.musl-libc.org/cgit/musl/commit/?id=725e17ed6dff4d0cd22487bb64470881e86a92e7

I noticed that some packages failed to compile due to "implicit declaration of baseline", so I added that declaration back for now.  At least elfutils  and rpcd-mod-luci failed. for that reason. In the longer run it might be better to fix the packages?   (cc @luizluca @jow- )


Example failure:

elfutils:
```
find-debuginfo.c: In function 'find_debuginfo_in_path':
find-debuginfo.c:167:58: error: implicit declaration of function 'basename' [-Werror=implicit-function-declaration]
  167 |   const char *file_basename = file_name == NULL ? NULL : basename (file_name);
      |                                                          ^~~~~~~~
find-debuginfo.c:167:56: error: pointer/integer type mismatch in conditional expression [-Werror]
  167 |   const char *file_basename = file_name == NULL ? NULL : basename (file_name);
      |                                                        ^
find-debuginfo.c:281:20: error: assignment to 'const char *' from 'int' makes pointer from integer without a cast [-Werror=int-conversion]
  281 |               file = basename (debuglink_file);
      |                    ^
find-debuginfo.c:309:32: error: passing argument 4 of 'try_open' makes pointer from integer without a cast [-Werror=int-conversion]
  309 |                                basename (file), &fname);
      |                                ^~~~~~~~~~~~~~~
      |                                |
      |                                int
find-debuginfo.c:44:60: note: expected 'const char *' but argument is of type 'int'
   44 |           const char *dir, const char *subdir, const char *debuglink,
      |                                                ~~~~~~~~~~~~^~~~~~~~~
cc1: all warnings being treated as errors
make[6]: *** [Makefile:717: find-debuginfo.o] Error 1
```

rpcd-mod-luci:

```
/Openwrt/DL-WRX36/build_dir/target-aarch64_cortex-a53_musl/rpcd-mod-luci-20230123/luci.c:676:53: error: implicit declaration of function 'basename' [-Werror=implicit-function-declaration]
  676 |                 blobmsg_add_string(&blob, "master", basename(link));
      |                                                     ^~~~~~~~
/Openwrt/DL-WRX36/build_dir/target-aarch64_cortex-a53_musl/rpcd-mod-luci-20230123/luci.c:676:53: error: passing argument 3 of 'blobmsg_add_string' makes pointer from integer without a cast [-Werror=int-conversion]
  676 |                 blobmsg_add_string(&blob, "master", basename(link));
      |                                                     ^~~~~~~~~~~~~~
      |                                                     |
      |                                                     int
In file included from /Openwrt/DL-WRX36/staging_dir/target-aarch64_cortex-a53_musl/usr/include/libubus.h:23,
                 from /Openwrt/DL-WRX36/build_dir/target-aarch64_cortex-a53_musl/rpcd-mod-luci-20230123/luci.c:43:
/Openwrt/DL-WRX36/staging_dir/target-aarch64_cortex-a53_musl/usr/include/libubox/blobmsg.h:235:72: note: expected 'const char *' but argument is of type 'int'
  235 | blobmsg_add_string(struct blob_buf *buf, const char *name, const char *string)
      |                                                            ~~~~~~~~~~~~^~~~~~
cc1: all warnings being treated as errors
ninja: build stopped: subcommand failed.
make[3]: *** [Makefile:58: /Openwrt/DL-WRX36/build_dir/target-aarch64_cortex-a53_musl/rpcd-mod-luci-20230123/.built] Error 1
```

There are likely also more packages failing similarly, but I noticed those two in my own build.

hnyman committed:
toolchain/musl: Update to 1.2.5

Update musl C library to 1.2.5

This release adds extension functions statx and preadv2/pwritev2,
with fallback implementations for older kernels, and adds two new
ports: loongarch64 and riscv32. Minor changes to the printf family
of functions have been made for conformance to new standards
interpretations/requirements. TLSDESC support for riscv64 has also
been added.

Bugs fixed include some DNS issues related to new TCP fallback
functionality, several rare race conditions, potentially incorrect
return value when glob aborts, and several signifiant arch-specific
bugs affecting TLSDESC on arm, riscv64 icache flushing, and
sh sigsetjmp and dlsym RTLD_NEXT.

1.2.5 release notes:

new features:
- statx function (linux extension; via syscall and fallback using fstatat)
- clone function is now usable and gives _Fork-like consistency in child
- statvfs now provides f_type result
- preadv2 and pwritev2 (linux extension) syscall wrappers
- riscv64 TLSDESC support

new ports:
- loongarch64
- riscv32

compatibility:
- DNS resolver can now handle answers with long CNAME chains
- string.h no longer provides (C23-incompat) non-prototype decl of basename
- fstatat statx backend now matches stat syscall non-automounting behavior
- mntent interfaces now handle escaped whitespace in paths/options

standards updates:
- printf %lc of nul wchar now produces output
- snprintf and swprintf no longer fail on n > INT_MAX
- ppoll is now exposed in default feature profile

bugs fixed:
- some long DNS answers were wrongly rejected despite new TCP support
- glob could wrongly return GLOB_NOMATCH if aborted before any matches
- multithreaded set*id could malfunction from thread sequencing logic bug
- certain use of threads after fork could deadlock thread-list lock
- posix_spawn child could deadlock in race with async parent death
- mbrtowc return value was wrong if argument n exceeded UINT_MAX
- 80-bit extended acoshl and powl got some corner cases wrong
- syslog incorrectly generated localized timestamps

arch-specific bugs fixed:
- arm (32-bit) TLSDESC malfunctioned due to addends being processed wrong
- riscv64 icache flush operation was non-functional
- sh sigsetjmp failed to properly restore call-saved register r8 on return
- sh dlsym RTLD_NEXT did not identify calling module correctly

Signed-off-by: Hannu Nyman <hannu.nyman@iki.fi>
[Removed patch adding basename hack again]
Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>

neheb commented:
Not a fan of that basename patch. I did a lot of work to remove local hacks from musl: https://github.com/openwrt/openwrt/commit/b519be1c522da3983277514a81dd83149737fbce

Failing packages should be fixed seperately.

hardfalcon commented:
The freshly released elfutils 0.191 actually contains a fix for this specific issue:
https://inbox.sourceware.org/elfutils-devel/CAJDtP-SJhJ8cjTpkNpi+F0nzVJ2pZsEai8Ewpp4yJmsO_H5-NA@mail.gmail.com/T/#u
https://sourceware.org/git/?p=elfutils.git;a=commit;h=a2194f6b305bf0d0b9dd49dccd0a5c21994c8eea

neheb commented:
that's somewhat interesting. Implies libgen.h inclusion is not quite correct...

guidosarducci committed:
elfutils: update to 1.91

Release Notes:
https://sourceware.org/pipermail/elfutils-devel/2024q1/006876.html

Refresh patch:
- 005-build_only_libs.patch

Signed-off-by: Tony Ambardar <itugrok@yahoo.com>

hackpascal commented:
Upgrading elfutils to 0.191 can definitly avoid the basename hack.

Ref: https://github.com/openwrt/openwrt/pull/14357/commits/a151c1a7aaa745f22df0165212e5178c6fd8da5f

Tested with loongarch64

hauke commented:
Most of the packages are compiling without the patch now. Thanks for the fixes.

elfutils needs an update, which is part of #14802.
utils/tini/ was not compiling, I will create a PR in the next days. 

I think it is fine if this is breaks some packages, the number is low and we can fix this after merging.

Could you please remove the musl patch and add this update for elfutils: https://github.com/openwrt/openwrt/pull/14802/commits/c1a56af82223147e0b29c9f8f5e608fdfda139c4

httpstorm commented:
@hauke 
I have not experimented with `musl-1.2.5`, though I was asked to test https://github.com/openwrt/openwrt/pull/14802 on macOS host. Both `coreutils` and `elfutils` fail to build for me. They depend on `gnulib` and fail if the wrong version is used, see https://github.com/openwrt/openwrt/pull/14802#issuecomment-2041055340. `coreutils` requires `gnulib-259829e7`. If we use that and make a few [changes](https://github.com/httpstorm/openwrt/tree/elfutils-0.191) (see the last 3 commits) to the `elfutils-0.191` patches from @guidosarducci and @mpratt14 [pr-14802-fix3](https://github.com/mpratt14/openwrt/tree/pr-14802-fix3), `coreutils` and `elfutils` build correctly.

Another [approach](https://github.com/httpstorm/openwrt/commit/2bfd756d278e4629a3753f3e46203136008e04a7) is reverting part of https://github.com/openwrt/openwrt/commit/09f8f21ae33402201c14a125a37b64746415642d, so that `coreutils` does not depend on the version of `gnulib`, but then we lose the ability to patch the `bootstrap` script. The pro is that 9.2-9.5 and potentially new releases work out of the box without `autoreconf` and we don't risk build errors due to mixing sources with incompatible versions. We may be able to patch `bootstrap` and pass `--no-bootstrap-sync`, though I have not tested this option. Currently `--bootstrap-sync` is used. @mpratt14 [suggested](https://github.com/openwrt/openwrt/pull/14802#issuecomment-2037901086) sending the `bootstrap` patch upstream. We should wait for his opinion.
```
  --bootstrap-sync         if this bootstrap script is not identical to
                           the version in the local gnulib sources,
                           update this script, and then restart it with
                           /bin/sh or the shell \$CONFIG_SHELL
  --no-bootstrap-sync      do not check whether bootstrap is out of sync
```

mpratt14 commented:
I have the rest of the fixes for the tools on macOS **most likely** ready, just waiting for your testing @httpstorm as I don't have a Mac...

I had to rename functions and edit my system's headers to test it `¯\_(ツ)_/¯`

with this we should be able to be more flexible with the version of gnulib used.

if anyone else wants to take a look:
https://github.com/mpratt14/openwrt/commits/pr-14802-fix-final/

hauke commented:
I updated the PR.
I updated elfutils from version 0.189 to version 0.191 in the packages directory and kept elfutils in tools at the old version 0.189 in this PR. 
As far as I understood elfutils in tools does not compile on MacOS in version 0.191.

I removed the basename hack patch from the commit. In addition rtty also does not compile without it.

mpratt14 commented:
@hauke Georgi and I have been working on the macOS fixes for elfutils in tools, we're almost done

it should probably be handled in the other PR anyway...

httpstorm commented:
@hauke 
I remember successfully building `packages/utils/coreutils-9.5` on macOS, though I have not deployed any images yet. I'll test it again once we finish working on the tools with Michael. We can update it. The rest of the packages I use build fine on macOS, except for the current version of `packages/lang/perl-5.38.2` which fails, so I am using `5.28.1`.

hauke commented:
I created PRs for some broken packages after the basename include change:
https://github.com/openwrt/packages/pull/23929
https://github.com/openwrt/packages/pull/23930
https://github.com/openwrt/packages/pull/23931
https://github.com/openwrt/packages/pull/23932
